File: path_sum.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a boolean
    def hasPathSum(self, root, sum):
        if root == None:    return False
        re = self.dfs(root, 0, [])
        return sum in re
        
    def dfs(self, node, s, res):
        s = s + node.val
        if node.left == None and node.right == None:
            res.append(s)
        if node.left != None:
            self.dfs(node.left, s, res)
        if node.right != None:
            self.dfs(node.right, s, res)
        return res

File: Leetcode_Java/PathSum.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) {
        val = x;
    }
}

public class PathSum {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }
        if (root.left == null && root.right == null) {
            return root.val == sum;
        }
        return hasPathSum(root.left, sum - root.val)
            || hasPathSum(root.right, sum - root.val);
    }

    public static void main(String[] args) {
        PathSum test = new PathSum();
        TreeNode one = new TreeNode(1);
        TreeNode two = new TreeNode(2);
        TreeNode three = new TreeNode(3);
        one.left = two;
        one.right = three;
        System.out.println(test.hasPathSum(one, 5));
    }
}